File: Sources/PiDeskKit/Models/Interaction.swift
import Foundation

/// Pi's `extension_ui_request` method. Tolerant, because a future Pi can introduce a dialog kind
/// this build has never seen and a remote client must still show *something* rather than
/// silently stranding a run that is blocked waiting for an answer.
public struct InteractionMethod: TolerantRawRepresentable {
    public let rawValue: String
    public init(rawValue: String) { self.rawValue = rawValue }

    public static let select = InteractionMethod(rawValue: "select")
    public static let confirm = InteractionMethod(rawValue: "confirm")
    public static let input = InteractionMethod(rawValue: "input")
    public static let editor = InteractionMethod(rawValue: "editor")

    public static var knownCases: [InteractionMethod] { [.select, .confirm, .input, .editor] }
    public static func other(_ rawValue: String) -> InteractionMethod { InteractionMethod(rawValue: rawValue) }

    /// A remote client can render an answer UI for these. Anything else must degrade to a
    /// visible "answer this on the Mac" card whose only safe action is Cancel.
    public var isAnswerable: Bool { Self.knownCases.contains(self) }
}

/// One option of a matched `ask_user_question` question. The raw string a response must carry
/// (`value`) is kept separate from the human label so a client never has to reconstruct Pi's own
/// option encoding.
public struct InteractionOption: Codable, Hashable, Sendable, Identifiable {
    public var id: Int
    public var value: String
    public var label: String
    public var description: String?
    public var preview: String?

    public init(id: Int, value: String, label: String, description: String? = nil, preview: String? = nil) {
        self.id = id
        self.value = value
        self.label = label
        self.description = description
        self.preview = preview
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(Int.self, forKey: .id) ?? 0
        value = try container.decodeIfPresent(String.self, forKey: .value) ?? ""
        label = try container.decodeIfPresent(String.self, forKey: .label) ?? value
        description = try container.decodeIfPresent(String.self, forKey: .description)
        preview = try container.decodeIfPresent(String.self, forKey: .preview)
    }
}

/// A dialog a daemon-owned Pi run is *currently blocked on*. It exists only while the run waits:
/// once answered, cancelled, or expired it is published one last time with `resolvedAt` set and
/// then forgotten. Nothing is ever answered automatically — an expiry sends Pi an explicit
/// cancellation so the run can unwind instead of burning its whole timeout.
public struct PendingInteraction: Codable, Hashable, Sendable, Identifiable {
    public var id: String
    public var runId: String
    public var threadId: String?
    public var method: InteractionMethod
    public var title: String
    public var message: String?
    /// The exact strings Pi offered, in order; a `select` response must be one of these.
    public var options: [String]
    public var placeholder: String?
    public var prefill: String?
    public var createdAt: Date
    public var expiresAt: Date
    public var resolvedAt: Date?

    // Enrichment from a matched `ask_user_question` tool call in the same run. Absent for a
    // plain approval/permission dialog, and absent whenever the match is not certain.
    public var header: String?
    public var multiSelect: Bool
    public var choices: [InteractionOption]
    public var questionIndex: Int?
    public var questionCount: Int?

    public init(
        id: String,
        runId: String,
        threadId: String? = nil,
        method: InteractionMethod,
        title: String,
        message: String? = nil,
        options: [String] = [],
        placeholder: String? = nil,
        prefill: String? = nil,
        createdAt: Date = Date(),
        expiresAt: Date,
        resolvedAt: Date? = nil,
        header: String? = nil,
        multiSelect: Bool = false,
        choices: [InteractionOption] = [],
        questionIndex: Int? = nil,
        questionCount: Int? = nil
    ) {
        self.id = id
        self.runId = runId
        self.threadId = threadId
        self.method = method
        self.title = title
        self.message = message
        self.options = options
        self.placeholder = placeholder
        self.prefill = prefill
        self.createdAt = createdAt
        self.expiresAt = expiresAt
        self.resolvedAt = resolvedAt
        self.header = header
        self.multiSelect = multiSelect
        self.choices = choices
        self.questionIndex = questionIndex
        self.questionCount = questionCount
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        runId = try container.decodeIfPresent(String.self, forKey: .runId) ?? ""
        threadId = try container.decodeIfPresent(String.self, forKey: .threadId)
        method = try container.decodeIfPresent(InteractionMethod.self, forKey: .method) ?? .other("unknown")
        title = try container.decodeIfPresent(String.self, forKey: .title) ?? "Pi"
        message = try container.decodeIfPresent(String.self, forKey: .message)
        options = try container.decodeIfPresent([String].self, forKey: .options) ?? []
        placeholder = try container.decodeIfPresent(String.self, forKey: .placeholder)
        prefill = try container.decodeIfPresent(String.self, forKey: .prefill)
        createdAt = try container.decodeIfPresent(Date.self, forKey: .createdAt) ?? .distantPast
        expiresAt = try container.decodeIfPresent(Date.self, forKey: .expiresAt) ?? .distantFuture
        resolvedAt = try container.decodeIfPresent(Date.self, forKey: .resolvedAt)
        header = try container.decodeIfPresent(String.self, forKey: .header)
        multiSelect = try container.decodeIfPresent(Bool.self, forKey: .multiSelect) ?? false
        choices = try container.decodeIfPresent([InteractionOption].self, forKey: .choices) ?? []
        questionIndex = try container.decodeIfPresent(Int.self, forKey: .questionIndex)
        questionCount = try container.decodeIfPresent(Int.self, forKey: .questionCount)
    }
}

public struct InteractionListResponse: Codable, Sendable {
    public var interactions: [PendingInteraction]
    public init(interactions: [PendingInteraction]) { self.interactions = interactions }
}

/// Exactly one of `value`, `confirmed`, or `cancelled` decides the answer, matching the app's own
/// `extension_ui_response` shape. An empty body is a cancellation, never a blank answer.
public struct InteractionRespondRequest: Codable, Sendable {
    public var value: String?
    public var confirmed: Bool?
    public var cancelled: Bool?

    public init(value: String? = nil, confirmed: Bool? = nil, cancelled: Bool? = nil) {
        self.value = value
        self.confirmed = confirmed
        self.cancelled = cancelled
    }
}

public struct InteractionRespondResponse: Codable, Sendable {
    public var accepted: Bool
    public init(accepted: Bool) { self.accepted = accepted }
}

File: Sources/PiDeskWeb/Site/js/views/interaction.js
import { h, mount } from "../dom.js";

// One dialog a daemon run is currently blocked on: an `ask_user_question` step, a permission
// prompt, or anything else Pi asked for over `extension_ui_request`.
//
// Two rules shape everything here:
//   1. Nothing is ever answered on the reader's behalf. There is no default selection, no
//      auto-submit, and no implicit "OK"; the only actions are ones a person took.
//   2. A dialog this build cannot render must still be *visible and escapable*. Pi is blocked
//      until it gets an answer, so an unknown method or an empty option list degrades to a card
//      that says so and offers Cancel, never to a silently skipped request.
//
// Values are submitted exactly as Pi offered them (`choices[].value`), so no option encoding is
// reconstructed on this side: a single-select answer is the raw option string, and a multi-select
// answer is the comma-separated 1-based index list the questionnaire plugin reads.

const ANSWERABLE = new Set(["select", "confirm", "input", "editor"]);

export function renderInteraction(interaction, { respond, describeError }) {
  const errorEl = h("div", { class: "inline-error", role: "alert", hidden: true });
  const body = h("div", { class: "interaction-body" });
  const actionsEl = h("div", { class: "interaction-actions" });

  const step =
    Number.isInteger(interaction.questionIndex) && Number.isInteger(interaction.questionCount)
      ? `Question ${interaction.questionIndex + 1} of ${interaction.questionCount}`
      : null;

  const node = h(
    "section",
    { class: "interaction", "aria-label": interaction.title || "Pi is asking" },
    h(
      "header",
      { class: "interaction-head" },
      h("span", { class: "interaction-chip" }, interaction.header || "Pi needs an answer"),
      step ? h("span", { class: "interaction-step" }, step) : null
    ),
    h("h2", { class: "interaction-title" }, interaction.title || "Pi is asking"),
    interaction.message ? h("p", { class: "interaction-message" }, interaction.message) : null,
    body,
    errorEl,
    actionsEl
  );

  let busy = false;

  function send(payload) {
    if (busy) return;
    busy = true;
    errorEl.hidden = true;
    node.setAttribute("aria-busy", "true");
    for (const control of node.querySelectorAll("button, input, textarea")) control.disabled = true;
    respond(payload).catch((err) => {
      busy = false;
      node.removeAttribute("aria-busy");
      for (const control of node.querySelectorAll("button, input, textarea")) control.disabled = false;
      errorEl.hidden = false;
      errorEl.textContent = describeError(err);
    });
  }

  const cancelBtn = h("button", { class: "btn", type: "button", onclick: () => send({ cancelled: true }) }, "Cancel");
  const method = interaction.method;
  const choices = Array.isArray(interaction.choices) && interaction.choices.length
    ? interaction.choices
    : (interaction.options || []).map((value, index) => ({ id: index, value, label: value }));

  if (method === "confirm") {
    mount(body, h("p", { class: "interaction-hint" }, "Answer on this device to let the run continue."));
    mount(actionsEl, [
      h("button", { class: "btn btn-primary", type: "button", onclick: () => send({ confirmed: true }) }, "Yes"),
      h("button", { class: "btn", type: "button", onclick: () => send({ confirmed: false }) }, "No"),
      cancelBtn
    ]);
  } else if (method === "select" && choices.length) {
    // A radio group, not a list of buttons: one tap must select, a second must submit, so a
    // mis-tap on a phone is recoverable instead of instantly answering for the reader.
    const name = `opt-${interaction.id}`;
    const submitBtn = h("button", { class: "btn btn-primary", type: "submit", disabled: true }, "Submit");
    const form = h(
      "form",
      {
        class: "interaction-options",
        role: "radiogroup",
        "aria-label": interaction.title || "Options",
        onsubmit: (event) => {
          event.preventDefault();
          const checked = form.querySelector("input:checked");
          if (checked) send({ value: checked.value });
        },
        onchange: () => {
          submitBtn.disabled = !form.querySelector("input:checked");
        }
      },
      choices.map((choice) => optionRow({ type: "radio", name, choice })),
      h("div", { class: "interaction-actions" }, submitBtn, cancelBtn)
    );
    mount(body, form);
  } else if (method === "input" && interaction.multiSelect && choices.length) {
    // The questionnaire plugin models a multi-select question as a typed `input` whose value is a
    // comma-separated 1-based index list. Checkboxes produce that string, so the reader never has
    // to know the encoding — and the free-text field below stays available for "none of these"
    // or an answer that is not on the list.
    const name = `opt-${interaction.id}`;
    const custom = h("input", {
      type: "text",
      class: "interaction-custom",
      "aria-label": "Type your own answer instead",
      placeholder: interaction.placeholder || "Or type your own answer"
    });
    const form = h(
      "form",
      {
        class: "interaction-options",
        "aria-label": interaction.title || "Options",
        onsubmit: (event) => {
          event.preventDefault();
          const typed = custom.value.trim();
          if (typed) return send({ value: typed });
          const selected = [...form.querySelectorAll("input[type=checkbox]:checked")].map((input) => input.value);
          // An empty multi-select is a real answer ("none of these"), which the plugin reads as
          // an empty input; it is the reader's choice, not a missing one.
          send({ value: selected.join(",") });
        }
      },
      choices.map((choice) => optionRow({ type: "checkbox", name, choice })),
      custom,
      h(
        "div",
        { class: "interaction-actions" },
        h("button", { class: "btn btn-primary", type: "submit" }, "Submit"),
        cancelBtn
      )
    );
    mount(body, form);
  } else if (method === "input" || method === "editor") {
    const field =
      method === "editor"
        ? h("textarea", { class: "interaction-text", rows: "6", "aria-label": interaction.title || "Answer" })
        : h("input", { type: "text", class: "interaction-text", "aria-label": interaction.title || "Answer" });
    if (interaction.placeholder) field.setAttribute("placeholder", interaction.placeholder);
    if (interaction.prefill) field.value = interaction.prefill;
    const form = h(
      "form",
      {
        class: "interaction-form",
        onsubmit: (event) => {
          event.preventDefault();
          send({ value: field.value });
        }
      },
      field,
      h(
        "div",
        { class: "interaction-actions" },
        h("button", { class: "btn btn-primary", type: "submit" }, "Submit"),
        cancelBtn
      )
    );
    mount(body, form);
  } else {
    // Either a method this build has never seen, or a known one whose payload cannot be answered
    // here (a `select` with no options). Say so plainly and leave exactly one safe way out.
    mount(
      body,
      h(
        "p",
        { class: "interaction-hint" },
        ANSWERABLE.has(method)
          ? "This prompt arrived without any options, so it cannot be answered here. Open the thread on your Mac, or cancel it."
          : "This kind of prompt needs the Mac app. Answer it there, or cancel it to let the run continue."
      )
    );
    mount(actionsEl, cancelBtn);
  }

  return node;
}

function optionRow({ type, name, choice }) {
  const input = h("input", { type, name, value: choice.value });
  return h(
    "label",
    { class: "interaction-option" },
    input,
    h(
      "span",
      { class: "interaction-option-text" },
      h("span", { class: "interaction-option-label" }, choice.label),
      choice.description ? h("span", { class: "interaction-option-desc" }, choice.description) : null,
      choice.preview ? h("pre", { class: "interaction-option-preview" }, choice.preview) : null
    )
  );
}
